packages/compiler/core/parser.ts

```diff
@@ -255,6 +255,17 @@ namespace ListKind {
 }
 
 export function parse(code: string | SourceFile, options: ParseOptions = {}): CadlScriptNode {
+  const parser = createParser(code, options);
+  return parser.parseCadlScript();
+}
+
+export interface Parser {
+  parseDiagnostics: Diagnostic[];
+  parseCadlScript(): CadlScriptNode;
+  parseStandaloneReferenceExpression(): TypeReferenceNode;
+}
+
+export function createParser(code: string | SourceFile, options: ParseOptions = {}): Parser {
```

Comment:
I wonder if the state here is sufficiently reset to call parseXxx more than once.